File: zskiplist.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

const (
	ZSKIPLIST_MAX_LEVEL = 32
	ZSKIPLIST_P         = 0.25
)

type zskiplistLevel struct {
	Forward *zskiplistNode
	Span    int //节点与节点间的跨度
}

type zskiplistNode struct {
	Key   int
	Value int
	Level []*zskiplistLevel
}

type zskiplist struct {
	Header *zskiplistNode
	Level  int
}

func (zslNode *zskiplistNode) equal(node *zskiplistNode) bool {
	if node == nil {
		return false
	}
	return zslNode.Key == node.Key && zslNode.Value == node.Value
}

func zslCreateNode(key int, value int) *zskiplistNode {
	node := &zskiplistNode{Key: key, Value: value, Level: make([]*zskiplistLevel, ZSKIPLIST_MAX_LEVEL)}
	for i := 0; i < ZSKIPLIST_MAX_LEVEL; i++ {
		node.Level[i] = &zskiplistLevel{Span: 0}
	}
	return node
}

func zslCreate() *zskiplist {
	return &zskiplist{Level: 0, Header: zslCreateNode(0, 0)}
}

func zslRandomLevel() int {
	ran := rand.New(rand.NewSource(time.Now().UnixNano()))
	level := ran.Intn(ZSKIPLIST_MAX_LEVEL)
	if level <= 0 {
		level = 1
	}
	return level
}

func (zsl *zskiplist) zslInsert(key int, value int) *zskiplistNode {
	update := make([]*zskiplistNode, ZSKIPLIST_MAX_LEVEL)
	header := zsl.Header
	//lookup position to insert
	for i := zsl.Level - 1; i >= 0; i-- {
		for header.Level[i].Forward != nil {
			if header.Level[i].Forward.Key < key ||
				(header.Level[i].Forward.Key == key && header.Level[i].Forward.Value < value) {
				header = header.Level[i].Forward // insert here
			} else {
				break
			}
		}
		update[i] = header
	}
	//locate at which level
	level := zslRandomLevel()
	//lastLevel := zsl.Level
	if level > zsl.Level {
		for i := zsl.Level; i < level; i++ {
			update[i] = zsl.Header
		}
		zsl.Level = level //max level in used
	}
	//insert data node
	node := zslCreateNode(key, value)
	for i := 0; i < level; i++ {
		node.Level[i].Forward = update[i].Level[i].Forward
		update[i].Level[i].Forward = node
		update[i].Level[i].Span = 1

	}
	return node
}

func (zsl *zskiplist) zslDeleteNode(node *zskiplistNode, update []*zskiplistNode) {
	for i := 0; i < zsl.Level; i++ {
		if update[i].Level[i].Forward != nil && update[i].Level[i].Forward.equal(node) {
			update[i].Level[i].Forward = node.Level[i].Forward
		}
	}
	for zsl.Level > 1 && zsl.Header.Level[zsl.Level-1].Forward == nil {
		zsl.Level -= 1
	}
}

func (zsl *zskiplist) zslDelete(key int, value int) bool {
	update := make([]*zskiplistNode, ZSKIPLIST_MAX_LEVEL)
	header := zsl.Header
	for i := zsl.Level - 1; i >= 0; i-- {
		for header.Level[i].Forward != nil {
			if header.Level[i].Forward.Key < key || (header.Level[i].Forward.Key == key &&
				header.Level[i].Forward.Value < value) {
				header = header.Level[i].Forward
			} else {
				break
			}
		}
		update[i] = header
	}

	node := header.Level[0].Forward
	if node != nil && key == node.Key && value == node.Value {
		zsl.zslDeleteNode(node, update)
		return true
	} else {
		return false
	}
}

func (zsl *zskiplist) zslPrint() {
	header := zsl.Header
	for i := zsl.Level - 1; i >= 0; i-- {
		p := header.Level[i].Forward
		for p != nil {
			fmt.Printf(" %d|%d", p.Key, p.Value)
			p = p.Level[i].Forward
		}
		fmt.Println()
	}
}

func main() {
	zslist := zslCreate()
	zslist.zslInsert(1, 2)
	zslist.zslInsert(4, 5)
	//zslist.zslInsert(2, 2)
	//zslist.zslPrint()
	//zslist.zslInsert(7, 2)
	//zslist.zslInsert(7, 3)
	//zslist.zslInsert(7, 3)
	//zslist.zslPrint()
	//ret := zslist.zslDelete(7, 3)
	//fmt.Println(ret)
	zslist.zslPrint()
}

File: zskiplist.c
#include<stdio.h>
#include<stdlib.h>
 
#define ZSKIPLIST_MAXLEVEL 32
#define ZSKIPLIST_P 0.25
#include <math.h>
 
//跳表节点
typedef struct zskiplistNode {
    int key;
    int value;
    struct zskiplistLevel {
        struct zskiplistNode *forward;
    } level[1];
} zskiplistNode;
 
//跳表
typedef struct zskiplist {
    struct zskiplistNode *header;
    int level;
} zskiplist;

//创建跳表的节点
zskiplistNode *zslCreateNode(int level, int key, int value) {
    zskiplistNode *zn = (zskiplistNode *)malloc(sizeof(*zn)+level*sizeof(zn->level));
    zn->key = key;
    zn->value = value;
    return zn;
}

//初始化跳表
zskiplist *zslCreate(void) {
    int j;
    zskiplist *zsl;
    zsl = (zskiplist *) malloc(sizeof(*zsl));
    zsl->level = 1;//将层级设置为1
    zsl->header = zslCreateNode(ZSKIPLIST_MAXLEVEL,NULL,NULL);
    for (j = 0; j < ZSKIPLIST_MAXLEVEL; j++) {
        zsl->header->level[j].forward = NULL;
    }
    return zsl;
}

//向跳表中插入元素时，随机一个层级，表示插入在哪一层
int zslRandomLevel(void) {
    int level = 1;
    while ((rand()&0xFFFF) < (ZSKIPLIST_P * 0xFFFF))
        level += 1;
    return (level<ZSKIPLIST_MAXLEVEL) ? level : ZSKIPLIST_MAXLEVEL;
}

//向跳表中插入元素
zskiplistNode *zslInsert(zskiplist *zsl, int key, int value) {
    zskiplistNode *update[ZSKIPLIST_MAXLEVEL], *x;
    int i, level;
    x = zsl->header;
    //在跳表中寻找合适的位置并插入元素
    for (i = zsl->level-1; i >= 0; i--) {
        while (x->level[i].forward &&
            (x->level[i].forward->key < key ||
                (x->level[i].forward->key == key &&
                x->level[i].forward->value < value))) {
            x = x->level[i].forward;
        }
        update[i] = x;
    }
    //获取元素所在的随机层数
    level = zslRandomLevel();
    if (level > zsl->level) {
        for (i = zsl->level; i < level; i++) {
            update[i] = zsl->header;
        }
        zsl->level = level;
    }
    //为新创建的元素创建数据节点
    x = zslCreateNode(level,key,value);
    for (i = 0; i < level; i++) {
        x->level[i].forward = update[i]->level[i].forward;
        update[i]->level[i].forward = x;
    }
    return x;
}

//跳表中删除节点的操作
void zslDeleteNode(zskiplist *zsl, zskiplistNode *x, zskiplistNode **update) {
    int i;
    for (i = 0; i < zsl->level; i++) {
        if (update[i]->level[i].forward == x) {
            update[i]->level[i].forward = x->level[i].forward;
        }
    }
    //如果层数变了，相应的将层数进行减1操作
    while(zsl->level > 1 && zsl->header->level[zsl->level-1].forward == NULL)
        zsl->level--;
}

//从跳表中删除元素
int zslDelete(zskiplist *zsl, int key, int value) {
    zskiplistNode *update[ZSKIPLIST_MAXLEVEL], *x;
    int i;
    x = zsl->header;
    //寻找待删除元素
    for (i = zsl->level-1; i >= 0; i--) {
        while (x->level[i].forward &&
            (x->level[i].forward->key < key ||
                (x->level[i].forward->key == key &&
                x->level[i].forward->value < value))) {
            x = x->level[i].forward;
        }
        update[i] = x;
    }
    x = x->level[0].forward;
    if (x && key == x->key && x->value == value) {
        zslDeleteNode(zsl, x, update);
        //别忘了释放节点所占用的存储空间
        free(x);
        return 1;
    } else {
        //未找到相应的元素
        return 0;
    }
    return 0;
}

//将链表中的元素打印出来
void printZslList(zskiplist *zsl) {
    zskiplistNode  *x;
    x = zsl->header;
    for (int i = zsl->level-1; i >= 0; i--) {
        zskiplistNode *p = x->level[i].forward;
        while (p) {
            printf(" %d|%d ",p->key,p->value);
            p = p->level[i].forward;
        }
        printf("\n");
    }
}

int main() {
    zskiplist *list = zslCreate();
    zslInsert(list,1,2);
    zslInsert(list,4,5);
    zslInsert(list,2,2);
    zslInsert(list,7,2);
    zslInsert(list,7,3);
    zslInsert(list,7,3);
    printZslList(list);
    //zslDelete(list,7,2);
    printZslList(list);
}
